### PR TITLE
ap-3836: enable allow listing

### PR DESCRIPTION
**Pull request by kmahern**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-3836)

- Enable alllow listing of MOJ ip addresses
- Add allow list to assure-hmrc-data-application-output (staging and production namespaces only).
- Decode and read this secret (escaping commas) and use it to set ingress annotations

Describe what you did and why.

## Checklist

Before you ask people to review this PR:

- Tests and linters should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: .helm/assure-hmrc-data/values/production.yaml [hunk 1]
```diff
@@ -22,6 +22,7 @@ ingress:
   annotations:
     external-dns.alpha.kubernetes.io/set-identifier: "assure-hmrc-data-laa-assure-hmrc-data-production-green"
     external-dns.alpha.kubernetes.io/aws-weight: "100"
+    nginx.ingress.kubernetes.io/whitelist-source-range: <%= ENV['ALLOW_LIST'] %>
```
Comment (jsugarman) [inline .helm/assure-hmrc-data/values/production.yaml:25]: i don't think the ENV var is available at this point?

File: .helm/assure-hmrc-data/values/uat.yaml [hunk 2]
```diff
@@ -23,6 +23,7 @@ ingress:
   annotations:
     external-dns.alpha.kubernetes.io/set-identifier: "assure-hmrc-data-laa-assure-hmrc-data-uat-green"
     external-dns.alpha.kubernetes.io/aws-weight: "100"
+    nginx.ingress.kubernetes.io/whitelist-source-range: <%= ENV['ALLOW_LIST'] %>
```
Comment (jsugarman) [inline .helm/assure-hmrc-data/values/uat.yaml:26]: we may not want to whitelist UAT (so we don't have to be on the VPN for it)

**Comment by jsugarman** [comment]
As per discussion there are a couple of options:

1. Store ALLOW_LIST as github action secret and amend the GHA workflow to set it.
```yaml
    - name: Helm deployment
      shell: bash
      env:
        ECR_URL: ${{ inputs.ecr-url }}
        GIT_SHA: ${{ github.sha }}
        KUBE_NAMESPACE: ${{ inputs.kube-namespace }}
        VALUES_FILE: .helm/assure-hmrc-data/values/${{ inputs.app-environment }}.yaml
        ALLOW_LIST: ${{ secrets.ALLOW_LIST }}
      run: |
        helm upgrade assure-hmrc-data .helm/assure-hmrc-data \
          --namespace ${KUBE_NAMESPACE} \
          --set image.repository="${ECR_URL}" \
          --set image.tag="${GIT_SHA}" \
          --set ingress.annotations."nginx.ingress.kubernetes.io/whitelist-source-range"="$ALLOW_LIST" \ <--- HERE
          --values ${VALUES_FILE} \
          --install \
          --wait
```
2. retrieve the secret value by script and set it in the helm command. This would require some parsing of the secrets to find the correct ones value, but has the advantage of making the kubernetes secret the source of truth. My example needs `jq` tool installed.
```shell
allow_list=$(kubectl --namespace ${KUBE_NAMESPACE} get secrets assure-hmrc-data-application-output -o json | jq '.data.allow_list | @base64d')

helm upgrade assure-hmrc-data .helm/assure-hmrc-data \
          --namespace ${KUBE_NAMESPACE} \
          --set image.repository="${ECR_URL}" \
          --set image.tag="${GIT_SHA}" \
          --set ingress.annotations."nginx.ingress.kubernetes.io/whitelist-source-range"="$allow_list" \
          --values ${VALUES_FILE} \
          --install \
          --wait
```

File: .helm/assure-hmrc-data/values/uat.yaml (hunk 2 above)
Comment (kmahern) [reply .helm/assure-hmrc-data/values/uat.yaml]: 👍 had added uat for testing purposes but now uat should not have allow listing in place

File: .helm/assure-hmrc-data/values/production.yaml (hunk 1 above)
Comment (kmahern) [reply .helm/assure-hmrc-data/values/production.yaml]: 👍

**Comment by kmahern** [comment]
Have followed the 2nd approach listed above, for consistency/in order to keep secrets in the same place (with a slight tweak to use jsonpath instead of jq).

File: .github/actions/deploy/action.yml [hunk 3]
```diff
@@ -48,10 +48,17 @@ runs:
         KUBE_NAMESPACE: ${{ inputs.kube-namespace }}
         VALUES_FILE: .helm/assure-hmrc-data/values/${{ inputs.app-environment }}.yaml
       run: |
+        echo "Decoding allow list"
+
+        allow_list=$(kubectl --namespace ${KUBE_NAMESPACE} get secrets assure-hmrc-data-application-output -o=jsonpath='{.data.allow_list}' | base64 --decode | sed "s/,/\\\,/g")
+
+        echo "Deploying commit: $GIT_SHA"
```
Comment (jsugarman) [inline .github/actions/deploy/action.yml:55]: I don't think we need this either?!

**Review by jsugarman — APPROVED** [review]
looks good, aside from the superfluos comments, IMO. 

I see you added the list to prod too 👍

**Comment by kmahern** [comment]
> looks good, aside from the superfluous comments, IMO.
> 
> I see you added the list to prod too 👍
> 
> **question:** did you try using `--set-string` to see if that would avoid the need to escape the commas

@jsugarman I get the same error when using --set-string

**Pull request merged**